daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/PState.scala

```diff
@@ -265,17 +276,29 @@ final class PState private (
   }
 
   def setVariable(vrd: VariableRuntimeData, newValue: DataValuePrimitive, referringContext: VariableRuntimeData, pstate: PState) {
-    this.setVariableMap(variableMap.setVariable(vrd, newValue, referringContext, pstate))
+    variableMap.setVariable(vrd, newValue, referringContext, pstate)
+    changedVariablesStack.top += vrd.globalQName
+  }
```

Comment:
Do we also need to modify changedVariableStack if a variable is read, since that does change the state of the variable? Or is resetting that handled somehow else?

Reply:
Good catch. I had missed this and it wasn't as easy as I expected to keep track of these variable reads. The code handling variable reads is in DState, which doesn't have direct access to things in PState, but I think I made a relatively clean fix for that.